src/main.py

```diff
@@ -73,76 +74,83 @@ async def query_metrics(execution_id: Optional[str], services: Optional[List[GCP
-            for disabled_project in disabled_projects:
-                projects_ids.remove(disabled_project)
-
-        setup_time = (time.time() - setup_start_time)
-        for project_id in projects_ids:
-            context.sfm[SfmKeys.setup_execution_time].update(project_id, setup_time)
-
-        context.start_processing_timestamp = time.time()
-
-        process_project_metrics_tasks = [
-            process_project_metrics(context, project_id, services, disabled_apis_by_project_id.get(project_id, set()))
-            for project_id
-            in projects_ids
-        ]
-        await asyncio.gather(*process_project_metrics_tasks, return_exceptions=True)
-        context.log(f"Fetched and pushed GCP data in {time.time() - context.start_processing_timestamp} s")
-
-        log_self_monitoring_metrics(context)
-        if context.self_monitoring_enabled:
-            await push_self_monitoring_metrics(context)
-
-        await gcp_session.close()
-        await dt_session.close()
+    gcp_session = init_gcp_client_session()
+    dt_session = init_dt_client_session()
+
+    setup_start_time = time.time()
+    token = await create_token(context, gcp_session)
+
+    if token is None:
+        context.log("Cannot proceed without authorization token, stopping the execution")
+        return
+    if not isinstance(token, str):
+        raise Exception(f"Failed to fetch access token, got non string value: {token}")
+
+    context.log("Successfully obtained access token")
+
+    project_id_owner = get_project_id_from_environment()
+
+    dynatrace_api_key = await fetch_dynatrace_api_key(gcp_session=gcp_session, project_id=project_id_owner, token=token)
+    dynatrace_url = await fetch_dynatrace_url(gcp_session=gcp_session, project_id=project_id_owner, token=token)
+    check_version(logging_context=context)
+    await check_dynatrace(logging_context=context,
+                          project_id=project_id_owner,
+                          dt_session=dt_session,
+                          dynatrace_url=dynatrace_url,
+                          dynatrace_access_key=dynatrace_api_key)
+
+    query_interval_min = get_query_interval_minutes()
+
+    print_metric_ingest_input = os.environ.get("PRINT_METRIC_INGEST_INPUT", "FALSE").upper() in ["TRUE", "YES"]
+    self_monitoring_enabled = os.environ.get('SELF_MONITORING_ENABLED', "FALSE").upper() in ["TRUE", "YES"]
+
+    context = MetricsContext(
+        gcp_session=gcp_session,
+        dt_session=dt_session,
+        project_id_owner=project_id_owner,
+        token=token,
+        execution_time=datetime.utcnow(),
+        execution_interval_seconds=60 * query_interval_min,
+        dynatrace_api_key=dynatrace_api_key,
+        dynatrace_url=dynatrace_url,
+        print_metric_ingest_input=print_metric_ingest_input,
+        self_monitoring_enabled=self_monitoring_enabled,
+        scheduled_execution_id=context.scheduled_execution_id
+    )
+
+    if context_provider.METRICS_CONTEXT is not None:
```

Comment:
In this big block, the only changes are
1) this 3 lines here about setting context_provider.METRICS_CONTEXT
2) removing "async with" for creating asyncio http client sessions, this "with" was useless anyhow it seems; removing it changed whole indentation in here so thats why big diff